packages/hardhat-ledger/src/provider.ts

```diff
@@ -0,0 +1,445 @@
+import { HardhatError } from "hardhat/internal/core/errors";
+import { ERRORS } from "hardhat/internal/core/errors-list";
+
+import { wrapTransport } from "./internal/wrap-transport";
+import * as cache from "./internal/cache";
+import { LedgerOptions, EthWrapper, Signature, Paths } from "./types";
+import {
+  HardhatLedgerConnectionError,
+  HardhatLedgerDerivationPathError,
+  HardhatLedgerError,
+  HardhatLedgerNotControlledAddressError,
+} from "./errors";
+
+export class LedgerProvider extends ProviderWrapperWithChainId {
+  public static readonly MAX_DERIVATION_ACCOUNTS = 20;
+  public static readonly DEFAULT_TIMEOUT = 3000;
+
+  public readonly paths: Paths = {}; // { address: path }
+  public name: string = "LedgerProvider";
+
+  protected _eth: EthWrapper | undefined;
+  private _isCreatingTransport = false;
+
+  public static async create(
+    options: LedgerOptions,
+    wrappedProvider: EIP1193Provider
+  ) {
+    const provider = new LedgerProvider(options, wrappedProvider);
+    await provider.init();
+    return provider;
+  }
+
+  constructor(
+    public readonly options: LedgerOptions,
+    protected readonly _wrappedProvider: EIP1193Provider
+  ) {
+    super(_wrappedProvider);
+
+    this.options.accounts = options.accounts.map((account) => {
+      if (!isValidAddress(account)) {
+        throw new HardhatLedgerError(
+          `The following ledger address from the config is invalid: ${account}`
+        );
+      }
+      return account.toLowerCase();
+    });
+  }
+
+  public get eth(): EthWrapper {
+    if (this._eth === undefined) {
+      throw new HardhatError(ERRORS.GENERAL.UNINITIALIZED_PROVIDER);
+    }
+    return this._eth;
+  }
+
+  public async init(): Promise<void> {
+    if (this._eth === undefined && !this._isCreatingTransport) {
+      this._isCreatingTransport = true;
+
+      const openTimeout =
+        this.options.openTimeout ?? LedgerProvider.DEFAULT_TIMEOUT;
+
+      const connectionTimeout =
+        this.options.connectionTimeout ?? LedgerProvider.DEFAULT_TIMEOUT;
+
+      try {
+        this.emit("connection_start");
+
+        const transport = await TransportNodeHid.create(
+          openTimeout,
+          connectionTimeout
+        );
+        this._eth = wrapTransport(transport);
+
+        this.emit("connection_success");
+      } catch (error) {
+        this.emit("connection_failure");
+
+        if (error instanceof Error) {
+          throw new HardhatLedgerConnectionError(error);
+        }
+
+        throw error;
+      }
+      this._isCreatingTransport = false;
+    }
+
+    try {
+      const paths = await cache.read<Paths>();
+      if (paths !== undefined) {
+        Object.assign(this.paths, paths);
+      }
+    } catch (error) {}
+  }
+
+  public async request(args: RequestArguments): Promise<unknown> {
+    const params = this._getParams(args);
+
+    if (
+      args.method === "eth_accounts" ||
+      args.method === "eth_requestAccounts"
```

Comment:
Hardhat doesn't support `eth_requestAccounts`, so I think we shouldn't add it here (otherwise it would be weird to have a new supported method after adding the plugin).

Reply:
Oh did not realize. Removed it, it'll error out with `MethodNotFoundError: Method eth_requestAccounts not found` by default, when forwarded to the inner provider

‼️ Just in case, `eth_requestAccounts` is supported on `LocalAccountsProvider` [here](https://github.com/NomicFoundation/hardhat/blob/main/packages/hardhat-core/src/internal/core/providers/accounts.ts#L51)